# Workshop Schedule

## Pre-reading

1. [Workflow (raw data to counts)](../lessons/01a_RNAseq_processing_workflow.md)
1. [Experimental design considerations](https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/experimental_planning_considerations.html)

## Day 1

| Time            |  Topic  | Instructor |
|:------------------------:|:------------------------------------------------:|:--------:|
| 10:00 - 10:30 | [Workshop Introduction](../lectures/Intro_to_workshop.pdf) | Radhika |
| 10:30 - 10:45 | [R refresher Q & A](../lessons/R_refresher.md) | Mary |
| 10:45 - 11:15 | RNA-seq pre-reading discussion | Meeta/Mary |
| 11:15 - 12:00 | [Intro to DGE / setting up DGE analysis](../lessons/01b_DGE_setup_and_overview.md) | Meeta |

### Self-Learning Part 1

1. [RNA-seq counts distribution](../lessons/01c_RNAseq_count_distribution.md)
1. [Count normalization](../lessons/02_DGE_count_normalization.md)
1. Sample-level QC (PCA and hierarchical clustering)

### Assignment #1
* All exercises from above lessons have been put together in [R script format](../homework/DGE_assignment_1.R)
* Dropbox link
* [Email us](mailto:hbctraining@hsph.harvard.edu) about questions regarding the homework that you need answered before the next class.
* Post questions that you would like to have reviewed in class [here](https://PollEv.com/hbctraining945).

---

## Day 2

| Time            |  Topic  | Instructor |
|:------------------------:|:------------------------------------------------:|:--------:|
| 10:00 - 11:00 | Self-learning lessons discussion | Radhika/Jihe |
| 11:00 - 11:30 | [Design formulas](../lessons/04a_design_formulas.md)  | Mary |
| 11:30 - 12:00 | [Hypothesis testing and multiple test correction](../lessons/05a_hypothesis_testing.md) | Meeta |

### Self-Learning Part 2
1. [Description of steps for DESeq2](../lessons/04b_DGE_DESeq2_analysis.md)
1. [Wald test results](../lessons/05b_wald_test_results.md)
1. [Summarizing results and extracting significant gene lists](../lessons/05c_summarizing_results.md)
1. Visualization

### Assignment #2
* R script
* Dropbox link
* [Email us](mailto:hbctraining@hsph.harvard.edu) about questions regarding the homework that you need answered before the next class.
* Post questions that you would like to have reviewed in class [here](https://PollEv.com/hbctraining945).

---

## Day 3

| Time            |  Topic  | Instructor |
|:------------------------:|:------------------------------------------------:|:--------:|
| 10:00 - 10:20 | Questions about self-learning lessons |  |
| 10:20 - 11:30 | In-class exercises |  |
| 11:30 - 12:00 | LRT until time course analysis |  |

### Self-Learning Part 3
1. Time course analysis
1. Gene annotation
1. Functional analysis

### Assignment #3
* R script
* Dropbox link
* [Email us](mailto:hbctraining@hsph.harvard.edu) about questions regarding the homework that you need answered before the next class.
* Post questions that you would like to have reviewed in class [here](https://PollEv.com/hbctraining945).

---

## Day 4

| Time            |  Topic  | Instructor |
|:------------------------:|:------------------------------------------------:|:--------:|
| 10:00 - 10:20 | Questions about self-learning lessons |  |
| 10:20 - 11:15 | In-class exercises |  |
| 11:15 - 11:45 | Discussion, Q & A |  |
| 11:45 - 12:00 | [Wrap Up](../lectures/R_nanocourse_wrapup_online.pdf) |  |

### Final Exercises
* Exercises

## Building on this workshop
* [Single-cell RNA-seq workshop](https://hbctraining.github.io/scRNA-seq/)
* [RMarkdown](https://hbctraining.github.io/Training-modules/Rmarkdown/)
* [Ggplot2 for functional analysis](https://hbctraining.github.io/Training-modules/Tidyverse_ggplot2/lessons/ggplot2.html)

## Resources
* [Online learning resources](https://hbctraining.github.io/bioinformatics_online/lists/online_trainings.html)
* [All hbctraining materials](https://hbctraining.github.io/main)
